## Problem 19: Counting Sundays

You are given the following information, but you may prefer to do some research for yourself.

- 1 Jan 1900 was a Monday.

- Thirty days has September,
    April, June and November.
    All the rest have thirty-one,
    Saving February alone,
    Which has twenty-eight, rain or shine.
    And on leap years, twenty-nine.

- A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.

How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

In [1]:
# python has datetime package
# Can return day of the week by giving date.
import datetime

In [2]:
''' 
Counts the number of a particular day falling on the first of a month between two dates.

eg:
>>> count_particular_day(1996, 5, 2023, 1, 'Monday')
>>> 45
'''

def count_particular_day(start_year : int, start_month : int, end_year : int, end_month : int, weekday : str) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Counts the number of a particular day falling on the first of a month between two dates.
        
        Uses the python library datetime to reutrn the weekday for a given date.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : start_year : The year to start the count from.
        : int : start_month : The month to start the count from.
        
        : int : end_year : The year to end the count from.
        : int : end_month : The month to end the count from.
        
        : str : weekday : The weekday to count.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : The number of that particular weekday between the dates given.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> count_particular_day(1996, 5, 2023, 1, 'Monday')
        >>> 45
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(start_year, int): raise ValueError('Please enter an integer > 0 for the start_year argument.')
    if start_year < 1: raise ValueError('Please enter an integer > 0 for the start_year argument.')
    if not isinstance(start_month, int): raise ValueError('Please enter an integer > 0 for the start_month argument.')
    if start_month < 1: raise ValueError('Please enter an integer > 0 for the start_month argument.')
    
    if not isinstance(end_year, int): raise ValueError('Please enter an integer > 0 for the end_year argument.')
    if end_year < 1: raise ValueError('Please enter an integer > 0 for the end_year argument.')
    if not isinstance(end_month, int): raise ValueError('Please enter an integer > 0 for the end_month argument.')
    if end_month < 1: raise ValueError('Please enter an integer > 0 for the end_month argument.')
    
    if weekday == "Monday":
        weekday_index = 0
    elif weekday ==  "Tuesday":
        weekday_index = 1
    elif weekday ==  "Wednesday":
        weekday_index = 2
    elif weekday ==  "Thursday":
        weekday_index = 3
    elif weekday ==  "Friday":
        weekday_index = 4
    elif weekday ==  "Saturday":
        weekday_index = 5
    elif weekday ==  "Sunday":
        weekday_index = 6
    else:
        raise ValueError('Please enter a day of the week for the weekday argument. eg. Monday')
    
    weekday_count = 0
    
    # Loop over the years specified.
    for year in range(start_year, end_year+1):
        
        # Loop through the months, need to be careful about start and end months in the loop.
        
        # If start and end year are the same, then only count between start month and end month
        if (year == start_year) and (year == end_year):
            for month in range(start_month, end_month+1):
                if(datetime.date(year, month, 1).weekday() == weekday_index): weekday_count += 1
        
        # If we are on the start year, then start at the start month
        elif year == start_year:
            for month in range(start_month, 13):
                if(datetime.date(year, month, 1).weekday() == weekday_index): weekday_count += 1
        
        # If we are at the end year, then end at the end month
        elif year == end_year:
            for month in range(1, end_month+1):
                if(datetime.date(year, month, 1).weekday() == weekday_index): weekday_count += 1
        
        # If we are in a middle year, then count all months
        else:
            for month in range(1, 13):
                if(datetime.date(year, month, 1).weekday() == weekday_index): weekday_count += 1
    
    return weekday_count

In [3]:
solution = count_particular_day(1901, 1, 2000, 12, 'Sunday')

print(f'Between the dates of 01/01/1901 and 31/12/2000 inclusive, there were {solution} Sundays that fell on the first of the month.')

Between the dates of 01/01/1901 and 31/12/2000 inclusive, there were 171 Sundays that fell on the first of the month.


### Problem 19 Solution:

Between the dates of 01/01/1901 and 31/12/2000 inclusive, there were 171 Sundays that fell on the first of the month.